compiler/rustc_query_impl/src/plumbing.rs

```diff
@@ -151,25 +151,41 @@ impl QueryContext for QueryCtxt<'tcx> {
         diagnostics: Option<&Lock<ThinVec<Diagnostic>>>,
         compute: impl FnOnce() -> R,
     ) -> R {
-        // The `TyCtxt` stored in TLS has the same global interner lifetime
-        // as `self`, so we use `with_related_context` to relate the 'tcx lifetimes
-        // when accessing the `ImplicitCtxt`.
-        tls::with_related_context(**self, move |current_icx| {
-            // Update the `ImplicitCtxt` to point to our new query job.
-            let new_icx = ImplicitCtxt {
-                tcx: **self,
-                query: Some(token),
-                diagnostics,
-                layout_depth: current_icx.layout_depth,
-                task_deps: current_icx.task_deps,
-            };
+        unsafe {
+            let mut slot = std::mem::MaybeUninit::uninit();
+            let func = std::mem::MaybeUninit::new(compute);
```

Comment:
I am not really fond of the added unsafety for compile time improvements. The query system is already complicated enough.

Reply:
Yeah I'm inclined to say we either shouldn't accept this PR or accept the first version (https://github.com/rust-lang/rust/commit/309ef1400796324e4f7807d3627b05d4e8c5f48a).

Reply:
Is the `unsized_fn_params` feature complete enough to allow passing and calling a `dyn FnOnce`?

Reply:
I don't know, can try it later.